docs/spec/relayer/Packets.md

```diff
@@ -98,4 +115,82 @@ func CreateDatagram(ev SendPacketEvent,
 }    
 ```
 
+### WriteAcknowledgementEvent handler
 
+Successful handling of *WriteAcknowledgementEvent* leads to *PacketAcknowledgement* datagram creation.
+
+```golang
+func CreateDatagram(ev WriteAcknowledgementEvent, 
+                    chainA Chain,    // source chain
+                    chainB Chain,    // destination chain
+                    proofHeight Height) (PacketAcknowledgement, Error) {   
+
+    // Stage 1 
+    // Verify if acknowledment is committed to chain A and it is still pending
+    packetAck, packetAckCommitmentProof, error = 
+        GetPacketAcknowledgement(chainA, ev.port, ev.channel, ev.sequence, proofHeight) 
+    if error != nil { return (nil, error) }
+    
+    if packetAck == nil OR packetAck != hash(ev.acknowledgement) { 
+        // invalid event; bad provider
+        return (nil, Error.BADPROVIDER) 
+    }
+
+    // Stage 2 
+    // Execute checks IBC handler on chainB will execute
+    
+    // Fetch channelEnd from the chainA to be able to compute port and chain ids on destination chain
+    channelA, proof, error = GetChannel(chainA, ev.port, ev.channel, ev.height)
+    if error != nil { return (nil, error) }
+    
+    channelB, proof, error = 
+        GetChannel(chainB, channelA.counterpartyPortIdentifier, channelA.counterpartyChannelIdentifier, LATEST_HEIGHT)
+    if error != nil { return (nil, error) }
+    
+    if channelB == nil OR channel.state != OPEN  { (nil, Error.DROP) } 
+    // Note that we checked implicitly above that counterparty identifiers match each other
+               
```

Comment:
Are these channel retrievals and checks the same as for the SendPacket event above? If yes can we make a single function for this? If not maybe we cover the differences and why they are.

Reply:
Checks are similar, but error message is different. With recv packets we can retry on error (if connection and channel are not open)  while for acks we should drop event in this case. We can make a function for this (and we should at the code level) but at the spec level maybe it's better making it more verbose so reading it is easier.